File: ClientBankSWPython/controller/product_registration_controller.py
import sys
import requests
from controller.general_contoller import get_db

db = get_db()

def list_all_product_registration():
    registrations_ref = db.collection(u'products_registrations')
    docs = registrations_ref.stream()
    registrations = []
    for doc in docs:
        registrations.append(doc.to_dict())
    return registrations

def insert_product_registration(str_client_id, str_product_code, str_product_number, balance,
                                 registration_date, expiration_date, state):
    client_id = int(str_client_id)
    product_code = int(str_product_code)
    product_number = int(str_product_number)
    balance = float(balance)
    
    new_product_registration = {
        "clientId" : client_id,
        "productCode" : product_code,
        "productNumber" : product_number,
        "balance" : balance,
        "registrationDate" : registration_date,
        "expirationDate" : expiration_date,
        "state" : state
    }
    
    doc_ref = db.collection(u'products_registrations').document("{}-{}".format(str_client_id, str_product_code))
    doc_ref.set(new_product_registration)
    return True

def find_product_registration(str_client_id, str_product_code):
    doc_ref = db.collection(u'products_registrations').document("{}-{}".format(str_client_id, str_product_code))
    doc = doc_ref.get()
    if doc.exists:
        return doc.to_dict()
    else:
        raise Exception("Registro de producto no encontrado")

def update_product_registration(str_client_id, str_product_code, str_product_number, balance, 
                                registration_date, expiration_date, state):
    try:
        product_registration_to_update = find_product_registration(str_client_id, str_product_code)
    
        client_id = int(str_client_id)
        product_code = int(str_product_code)
        product_number = int(str_product_number)
        balance = float(balance)
        
        product_registration_to_update['clientId'] = client_id
        product_registration_to_update['productCode'] = product_code
        product_registration_to_update['productNumber'] = product_number
        product_registration_to_update['balance'] = balance
        product_registration_to_update['registrationDate'] = registration_date
        product_registration_to_update['expirationDate'] = expiration_date
        product_registration_to_update['state'] = state
        
        doc_ref = db.collection(u'products_registrations').document("{}-{}".format(str_client_id, str_product_code))
        doc_ref.set(product_registration_to_update)
        return True
    except Exception as ex:
        raise ex

def delete_product_registration(str_client_id, str_product_code):
    try:
        product_registration_to_delete = find_product_registration(str_client_id, str_product_code)
        db.collection(u'products_registrations').document("{}-{}".format(str_client_id, str_product_code)).delete
        return True
    except Exception as ex:
        raise ex

File: ClientBankSWNet/ClientBankSWNet/controller/ProductRegistrationController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Google.Cloud.Firestore;

namespace ClientBankSWNet.controller
{
    class ProductRegistrationController
    {
        private static ProductRegistrationController controller = null;
        private static FirestoreDb db;

        private ProductRegistrationController() 
        {
            db = GeneralController.DataBase;
        }

        public static ProductRegistrationController Instance
        {
            get
            {
                if (controller == null)
                    controller = new ProductRegistrationController();

                return controller;
            }
        }

        public async Task<List<Dictionary<string, object>>> ListAllProductRegistations()
        {
            List<Dictionary<string, object>> productRegistrations = new List<Dictionary<string, object>>();

            try
            {
                CollectionReference collRef = db.Collection("products_registrations");
                QuerySnapshot snapshot = await collRef.GetSnapshotAsync();

                foreach (DocumentSnapshot document in snapshot.Documents)
                {
                    Dictionary<string, object> documentDictionary = document.ToDictionary();
                    productRegistrations.Add(documentDictionary);
                }
            }
            catch (Exception ex)
            {
                throw ex;
            }

            return productRegistrations;
        }

        public async Task<bool> AddProductRegistration(String strClientId, int productCode, String strProductNumber,
                                                String strBalance, DateTime registrationDate, DateTime expirationDate,
                                                bool state)
        {
            if (strClientId.Equals("") || strProductNumber.Equals("") || strBalance.Equals("") ||
                registrationDate == null || expirationDate == null)
            {
                throw new Exception("Por favor diligenciar todos los campos");
            }
            else
            {
                int clientId = int.Parse(strClientId);
                int productNumber;
                double balance;

                try
                {
                    productNumber = int.Parse(strProductNumber);
                }
                catch (Exception)
                {
                    throw new Exception("Número de producto invalido.\nPor favor digite un valor numérico");
                }
                try
                {
                    balance = Double.Parse(strBalance);
                }
                catch (Exception)
                {
                    throw new Exception("Saldo invalido.\nPor favor digite un valor numérico");
                }

                Dictionary<string, object> newRegistration = new Dictionary<string, object>() {
                    {"clientId", clientId },
                    {"productCode" , productCode },
                    {"productNumber" , productNumber },
                    {"balance" , balance },
                    {"registrationDate" , registrationDate.ToString("yyyy-MM-dd")},
                    {"expirationDate" , expirationDate.ToString("yyyy-MM-dd")},
                    {"state" , state}
                };

                DocumentReference docRef = db.Collection("products_registrations").Document(String.Format("{0}-{1}", strClientId, productCode));
                await docRef.SetAsync(newRegistration);

                return true;
            }
        }

        public async Task<Dictionary<string, object>> FindProductRegistration(String strClientId, String strProductCode)
        {
            int clientId = int.Parse(strClientId);
            int productCode;
            try
            {
                productCode = int.Parse(strProductCode);
            }
            catch (Exception)
            {
                throw new Exception("Por favor seleccione un producto");
            }

            DocumentReference docRef = db.Collection("products_registrations").Document(String.Format("{0}-{1}", strClientId, productCode));
            DocumentSnapshot snapshot = await docRef.GetSnapshotAsync();

            if (snapshot.Exists)
            {
                Dictionary<string, object> productRegistration = snapshot.ToDictionary();
                return productRegistration;
            }
            else
            {
                throw new Exception("No se ha encontrado el registro del producto");
            }
        }

        public async Task<bool> UpdateProductRegistration(String strClientId, String strProductCode, String strProductNumber,
                                                String strBalance, DateTime registrationDate, DateTime expirationDate,
                                                bool state)
        {
            if (strProductNumber.Equals("") || strBalance.Equals("") || registrationDate == null || expirationDate == null)
            {
                throw new Exception("Por favor diligenciar todos los campos");
            }
            else
            {
                Dictionary<string, object> productRegistrationToUpdate;

                try
                {
                    productRegistrationToUpdate = await this.FindProductRegistration(strClientId, strProductCode);
                }
                catch (Exception ex)
                {
                    throw ex;
                }

                int clientId = int.Parse(strClientId);
                int productCode;
                int productNumber;
                double balance;

                try
                {
                    productCode = int.Parse(strProductCode);
                }
                catch (Exception)
                {
                    throw new Exception("Por favor seleccione un producto");
                }
                try
                {
                    productNumber = int.Parse(strProductNumber);
                }
                catch (Exception)
                {
                    throw new Exception("Número de producto invalido.\nPor favor digite un valor numérico");
                }
                try
                {
                    balance = Double.Parse(strBalance);
                }
                catch (Exception)
                {
                    throw new Exception("Saldo invalido.\nPor favor digite un valor numérico");
                }

                productRegistrationToUpdate = new Dictionary<string, object>() {
                    {"clientId", clientId },
                    {"productCode" , productCode },
                    {"productNumber" , productNumber },
                    {"balance" , balance },
                    {"registrationDate" , registrationDate.ToString("yyyy-MM-dd")},
                    {"expirationDate" , expirationDate.ToString("yyyy-MM-dd")},
                    {"state" , state}
                };

                DocumentReference docRef = db.Collection("products_registrations").Document(String.Format("{0}-{1}", strClientId, productCode));
                await docRef.SetAsync(productRegistrationToUpdate, SetOptions.MergeAll);

                return true;
            }

        }

        public async Task<bool> DeleteProductRegistration(String strClientId, String strProductCode)
        {
            Dictionary<string, object> productRegistrationToDelete;

            try
            {
                productRegistrationToDelete = await this.FindProductRegistration(strClientId, strProductCode);
            }
            catch (Exception ex)
            {
                throw ex;
            }

            DocumentReference docRef = db.Collection("products_registrations").Document(String.Format("{0}-{1}", strClientId, strProductCode));
            await docRef.DeleteAsync();

            return true;
        }
    }
}

